Added the clock target and required attributes

RameshIyyar opened the pull request.

RameshIyyar committed:
data: p10: Added the OSCREFCLK target

The OSCREFCLK target is required for the clock specific requirements
i.e. guard use case, maintain the system-specific information (i2c path,
clock target position) that needs to perform some action in the
firmware stack so added the clock target with common attributes
in this patch.

Attributes will be included based on the OSCREFCLK parent target.
Below are currently added attributes for oesrefclk target based on
its parent.
 - ATTR_FAPI_NAME
 - ATTR_HWAS_STATE
 - ATTR_PHYS_BIN_PATH
 - ATTR_PHYS_DEV_PATH
 - ATTR_TYPE

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>

RameshIyyar committed:
data: p10: Added the POSITION attribute

- Added the POSITION attribute for the OSCREFCLK target.

- The POSITION attribute value pulled from system MRW xml.
  The OSCREFCLK target POSITION attribute will contain the clock
  position and that needs to get to know the clock position.

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>

RameshIyyar commented:
Tested with Everest system MRW xml file. Clock target did not add into the Rainier system MRW XML file.

```
oscrefclk0 {
compatible = "ibm,power10-oscrefclk";
index = <0x00>;
ATTR_CLOCK_I2C_PATH = "/dev/i2c-0";
ATTR_FAPI_NAME = "NA";
ATTR_HWAS_STATE = [ 00 00 00 00 00 ];
ATTR_PHYS_BIN_PATH = [ 23 01 00 02 00 1A 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ];
ATTR_PHYS_DEV_PATH = "physical:sys-0/node-0/oscrefclk-0";
ATTR_POSITION = [ 00 01 ];
ATTR_TYPE = [ 1A ];
};

oscrefclk1 {
compatible = "ibm,power10-oscrefclk";
index = <0x01>;
ATTR_CLOCK_I2C_PATH = "/dev/i2c-1";
ATTR_FAPI_NAME = "NA";
ATTR_HWAS_STATE = [ 00 00 00 00 00 ];
ATTR_PHYS_BIN_PATH = [ 23 01 00 02 00 1A 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ];
ATTR_PHYS_DEV_PATH = "physical:sys-0/node-0/oscrefclk-1";
ATTR_POSITION = [ 00 00 ];
ATTR_TYPE = [ 1A ];
};
```

ojayanth commented on data/p10/filter_TargetsList.lsv:
```diff
@@ -31,3 +31,5 @@ unit-nx-power10
 # TPM target
 chip-NPCT750AAC
+# OSCREFCLK target
+chip-SI5332
```
Need @amitay input on pdbg side implementation.

RameshIyyar replied:
Sorry, I am not getting it, what is the impact on the pdbg side with this name? or do you mean in the general?

The target type is picked based on the system MRW XML (Rainier and Everest).

ojayanth replied:
You can ignore this comment , for got remove from here , pdbg_compatible_propMapping.lsv related changes. i place initiall her and forgot to remove.

ojayanth commented on attribute/attribute_target.c:
```diff
@@ -75,6 +75,7 @@ struct {
 
 	{ "TARGET_TYPE_TPM", "tpm", "tpm" },
 	{ "TARGET_TYPE_BMC", "bmc", "bmc" },
+	{ "TARGET_TYPE_OSCREFCLK", "oscrefclk", "oscrefclk" },
```
Commit message : Description looks contradictin , initiall mentioed fir guard purpose and the added additional attribute info. This target is not only use for guard purpose. Please update commit message accordingly

RameshIyyar replied:
Done

ojayanth commented on data/p10/pdbg_compatible_propMapping.lsv:
```diff
@@ -34,3 +34,5 @@ unit-occ-power10:ibm,power10-occ
 unit-thread:"ibm,power-thread", "ibm,power10-thread"
 # TPM target
 chip-NPCT750AAC:ibm,power10-tpm
+# OSCREFCLK target
```
Need @amitay input on pdbg side implementation.

RameshIyyar replied:
Yes, the pdbg target compatible property must be the same when adding the hardware unit on the pdbg side.

ojayanth commented on scripts/processMrw_bmc.pm:
```diff
@@ -50,6 +66,19 @@ sub buildBMCAffinity
+            $oscrefclk++;
+            $osrefclk_instance_per_node++;
+            # Hardcoding the sys instance as "0" since we will have only one system
+            $targetObj->{targeting}{SYS}[0]{NODES}[$node]{OSCREFCLK}[$osrefclk_instance_per_node]{KEY}= $target;
```
we will have only one node?

RameshIyyar replied:
we can have multi-node and that's why we are using the `$node` count in this path preparation. 

FYI, targets were read from MRW based on the order (same as FSP) so, I believe the order will be as expected in the multi-node system MRW but, if something is changed in the future then, we need to revisit based on the needs and currently we have only one node for p10 BMC system.

ojayanth replied:
your commnet "we will have only one system" should c"hange to we will have only one node"?

RameshIyyar replied:
sys and node are different right?
I think I meant "we will have only one sys target (system)" and the sys target can have more node target under that right?

RameshIyyar committed:
scripts: attribute: xml: Added notRequiredInDevTree tag support

- Added the "notRequiredInDevTree" tag support and that tag can be
  added into the BMC specific attribute definition xml.

- If the device tree and meta-data tools identified the added
  "notRequiredInDevTree" tag means, it won't consider that attribute
  to add in the generated device tree or meta-data files.

- This tag helps the auto-gen tool if tools need some attribute
  and that attribute value needs to be pulled from the system MRW xml.

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>

RameshIyyar committed:
scripts: genDTS: Added the OSCREFCLK target under the I2C target

- The application will perform the I2C read and write operation
  on the OSCREFCLK target to reset and to get to know the status.

- The libpdbg using the standard device tree address translation
  for the I2C operation so if the target needs an I2C operation
  to perform some action then, that target needs under
  the respective I2C instance target with its standard device tree
  address translation properties i.e address_cells, size_cells
  in the I2C target and the reg property needs to be present
  in the target which requires I2C operation.

- So, the OSCREFCLK target added under the respective I2C instance
  target by using the I2C bus connection (which is defined in the MRW)
  of the OSCREFCLK target. The I2C instance id and address values are
  pulled from the MRW I2C_PORT and I2C_ADDRESS attribute respectively
  by using that defined the I2C bus connection.

- Added the three custom properties (I2C_PORT, I2C_ADDRESS, and
  I2C_PARENT_PHYS_PATH) and that will be used by the pdata tools
  to form the target hierarchy as per pdbg expectation so if any target
  wants to perform the I2C operation then, that target can have
  these properties.

  - Note, these properties are added for tool purposes and application
    doesn't need that so those properties are not added in the generated
    device tree and meta-data files.

- For example, the OSCREFCLK target in the Everest device tree will be
  like below,

```
bmc0 {                         <-- I2C parent
    compatible = "bmc-ast2600";
    index = <0x00>;
    ...

    i2c-0 {                    <-- I2C Port
        #address-cells = <1>;
        #size-cells = <0>;

        oscrefclk0 {
            compatible = "ibm,power10-oscrefclk";
            reg = <0xD4>;      <-- I2C address
            index = <0x00>;
            ...
        };
    };

    i2c-1 {                    <-- I2C Port
        #address-cells = <1>;
        #size-cells = <0>;

        oscrefclk1 {
            compatible = "ibm,power10-oscrefclk";
            reg = <0xD4>;      <-- I2C address
            index = <0x01>;
            ...
        };
    };
};
```

Tested:
- Poweron the supported system with new generated device tree

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>

RameshIyyar commented:
Changed the OSCREFCLK target hierarchy as per pdbg expectation to perform the I2C read and write operation.

ojayanth reviewed (commented):
Looking good to me, Need some clock firmware experts apparoval before merge.

amitay commented on attribute/old/attribute_target.c:
```diff
@@ -75,6 +75,7 @@ struct {
 
 	{ "TARGET_TYPE_TPM", "tpm", "tpm" },
 	{ "TARGET_TYPE_BMC", "bmc", "bmc" },
+	{ "TARGET_TYPE_OSCREFCLK", "oscrefclk", "oscrefclk" },
```
This is the old attribute code.  The new code uses libdtree.  So you will have to modify `libdtree/dtree_util.c`.

RameshIyyar replied:
Thanks for pointing - I did not notice during rebase.

Added this line in `libdtree` and removed in the old code

RameshIyyar committed:
data: p10: Added clock target support for Rainier system

In the Rainier system mrw, the clock target id is different as per
board design so added that target id support in the meta-data
and tools to pull clock target from Rainier system mrw and add into
the cec device tree.

Tested:
```
bmc0 {                         <-- I2C parent
    compatible = "bmc-ast2600";
    index = <0x00>;
    ...

    i2c-8 {                    <-- I2C Port
        #address-cells = <1>;
        #size-cells = <0>;

        oscrefclk0 {
            compatible = "ibm,power10-oscrefclk";
            reg = <0xD4>;      <-- I2C address
            index = <0x00>;
            ...
        };
    };
};
```

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>

rajez133 reviewed (approved):
Device tree generated is working for clock firmware
